src/Raven.Server/Documents/PeriodicBackup/BackupTask.cs

```diff
@@ -613,36 +620,46 @@ private InternalBackupResult CreateLocalBackupOrSnapshot(PeriodicBackupStatus st
-                        var smugglerResult = _database.FullBackupTo(tempBackupFilePath, compressionLevel, excludeIndexes,
-                            info =>
-                            {
-                                AddInfo(info.Message);
 
-                                _backupResult.SnapshotBackup.ReadCount += info.FilesCount;
-                                if (sw.ElapsedMilliseconds > 0 && info.FilesCount > 0)
+                        using (var stream = GetStreamForBackupDestination(tempBackupFilePath, folderName, fileName))
+                        {
+                            var smugglerResult = _database.FullBackupTo(stream, compressionLevel, excludeIndexes,
+                                info =>
                                 {
-                                    AddInfo($"Backed up {_backupResult.SnapshotBackup.ReadCount} " +
-                                            $"file{(_backupResult.SnapshotBackup.ReadCount > 1 ? "s" : string.Empty)}");
-                                    sw.Restart();
-                                }
-                            }, TaskCancelToken.Token);
+                                    AddInfo(info.Message);
 
-                        EnsureSnapshotProcessed(databaseSummary, smugglerResult, indexesCount);
+                                    _backupResult.SnapshotBackup.ReadCount += info.FilesCount;
+                                    if (sw.ElapsedMilliseconds > 0 && info.FilesCount > 0)
+                                    {
+                                        AddInfo($"Backed up {_backupResult.SnapshotBackup.ReadCount} " +
+                                                $"file{(_backupResult.SnapshotBackup.ReadCount > 1 ? "s" : string.Empty)}");
+                                        sw.Restart();
+                                    }
+                                }, TaskCancelToken.Token);
+
+                            FlushToDiskIfNeeded(stream);
+
+                            EnsureSnapshotProcessed(databaseSummary, smugglerResult, indexesCount);
+                        }
 
                         AddInfo($"Backed up {_backupResult.SnapshotBackup.ReadCount} files, " +
                                 $"took: {totalSw.ElapsedMilliseconds:#,#;;0}ms");
                     }
 
-                    IOExtensions.RenameFile(tempBackupFilePath, backupFilePath);
+                    if (_directUploadDestination == DirectUploadDestination.Disabled)
+                        IOExtensions.RenameFile(tempBackupFilePath, backupFilePath);
 
                     status.LocalBackup.Exception = null;
                 }
                 catch (Exception e)
                 {
                     status.LocalBackup.Exception = e.ToString();
 
-                    // deleting the temp backup file if the backup failed
-                    DeleteFile(tempBackupFilePath);
+                    if (_directUploadDestination == DirectUploadDestination.Disabled)
```

Comment:
I wonder if we could handle it somehow better. Instead of doing `if / switch` and checking `DirectUploadDestination.Disabled` maybe we could introduce some abstraction.

Reply:
done